From 75ccddf52569485903726a764d4f9de041944b30 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Thu, 26 Oct 2023 11:41:38 -0500
Subject: [PATCH] Revert "Use tray icon from system theme only if option is
 set"

This reverts commit 77bd09bb8b3191e98cff7502284c657cf33e73f0.

This breaks the tray icon, see https://github.com/qbittorrent/qBittorrent/issues/19758
---
 src/gui/desktopintegration.cpp | 22 +++-------------------
 src/gui/desktopintegration.h   |  1 -
 src/gui/uithememanager.cpp     | 32 +++++++++++++++++++++++++++++++-
 src/gui/uithememanager.h       |  4 ++++
 4 files changed, 38 insertions(+), 21 deletions(-)

diff --git a/src/gui/desktopintegration.cpp b/src/gui/desktopintegration.cpp
index 2be2bc69d..069acd0b6 100644
--- a/src/gui/desktopintegration.cpp
+++ b/src/gui/desktopintegration.cpp
@@ -38,6 +38,7 @@
 #include <QSystemTrayIcon>
 #endif
 
+#include "base/logger.h"
 #include "base/preferences.h"
 #include "uithememanager.h"
 
@@ -241,7 +242,7 @@ void DesktopIntegration::onPreferencesChanged()
         if (m_systrayIcon)
         {
             // Reload systray icon
-            m_systrayIcon->setIcon(getSystrayIcon());
+            m_systrayIcon->setIcon(UIThemeManager::instance()->getSystrayIcon());
         }
         else
         {
@@ -262,7 +263,7 @@ void DesktopIntegration::createTrayIcon()
 {
     Q_ASSERT(!m_systrayIcon);
 
-    m_systrayIcon = new QSystemTrayIcon(getSystrayIcon(), this);
+    m_systrayIcon = new QSystemTrayIcon(UIThemeManager::instance()->getSystrayIcon(), this);
 
     m_systrayIcon->setToolTip(m_toolTip);
 
@@ -282,21 +283,4 @@ void DesktopIntegration::createTrayIcon()
     m_systrayIcon->show();
     emit stateChanged();
 }
-
-QIcon DesktopIntegration::getSystrayIcon() const
-{
-    const TrayIcon::Style style = Preferences::instance()->trayIconStyle();
-    switch (style)
-    {
-    default:
-    case TrayIcon::Style::Normal:
-        return UIThemeManager::instance()->getIcon(u"qbittorrent-tray"_s);
-
-    case TrayIcon::Style::MonoDark:
-        return UIThemeManager::instance()->getIcon(u"qbittorrent-tray-dark"_s);
-
-    case TrayIcon::Style::MonoLight:
-        return UIThemeManager::instance()->getIcon(u"qbittorrent-tray-light"_s);
-    }
-}
 #endif // Q_OS_MACOS
diff --git a/src/gui/desktopintegration.h b/src/gui/desktopintegration.h
index 62dbbc4a9..a103a7dd5 100644
--- a/src/gui/desktopintegration.h
+++ b/src/gui/desktopintegration.h
@@ -77,7 +77,6 @@ private:
     void onPreferencesChanged();
 #ifndef Q_OS_MACOS
     void createTrayIcon();
-    QIcon getSystrayIcon() const;
 #endif // Q_OS_MACOS
 
     CachedSettingValue<bool> m_storeNotificationEnabled;
diff --git a/src/gui/uithememanager.cpp b/src/gui/uithememanager.cpp
index aa49badd5..c974d7a25 100644
--- a/src/gui/uithememanager.cpp
+++ b/src/gui/uithememanager.cpp
@@ -121,7 +121,7 @@ QIcon UIThemeManager::getIcon(const QString &iconId, [[maybe_unused]] const QStr
     if (m_useSystemIcons)
     {
         auto icon = QIcon::fromTheme(iconId);
-        if (icon.isNull() || icon.availableSizes().isEmpty())
+        if (icon.name() != iconId)
             icon = QIcon::fromTheme(fallback, QIcon(m_themeSource->getIconPath(iconId, colorMode).data()));
         return icon;
     }
@@ -174,6 +174,36 @@ QColor UIThemeManager::getColor(const QString &id) const
     return color;
 }
 
+#ifndef Q_OS_MACOS
+QIcon UIThemeManager::getSystrayIcon() const
+{
+    const TrayIcon::Style style = Preferences::instance()->trayIconStyle();
+    switch (style)
+    {
+#if defined(Q_OS_UNIX)
+    case TrayIcon::Style::Normal:
+        return QIcon::fromTheme(u"qbittorrent-tray"_s);
+    case TrayIcon::Style::MonoDark:
+        return QIcon::fromTheme(u"qbittorrent-tray-dark"_s);
+    case TrayIcon::Style::MonoLight:
+        return QIcon::fromTheme(u"qbittorrent-tray-light"_s);
+#else
+    case TrayIcon::Style::Normal:
+        return getIcon(u"qbittorrent-tray"_s);
+    case TrayIcon::Style::MonoDark:
+        return getIcon(u"qbittorrent-tray-dark"_s);
+    case TrayIcon::Style::MonoLight:
+        return getIcon(u"qbittorrent-tray-light"_s);
+#endif
+    default:
+        break;
+    }
+
+    // As a failsafe in case the enum is invalid
+    return getIcon(u"qbittorrent-tray"_s);
+}
+#endif
+
 void UIThemeManager::applyPalette() const
 {
     struct ColorDescriptor
diff --git a/src/gui/uithememanager.h b/src/gui/uithememanager.h
index df5e0984c..3e182769d 100644
--- a/src/gui/uithememanager.h
+++ b/src/gui/uithememanager.h
@@ -57,6 +57,10 @@ public:
 
     QColor getColor(const QString &id) const;
 
+#ifndef Q_OS_MACOS
+    QIcon getSystrayIcon() const;
+#endif
+
 private:
     UIThemeManager(); // singleton class
 
-- 
2.42.0

